### PR TITLE
improved position based flow correction for lpe

### PR DESCRIPTION
**Pull request by jgoppert**
The main difference is that this treats flow as a position estimate in order to reduce long-term drift. It also has a high pass on the flow gyro which effectively removes gyro bias from the flow sensor without having to estimate it directly.

Flight test video, pretty good, mild wind, 5-10 mph, very cold :-/:
https://www.youtube.com/watch?v=0PqPsRMFyhw

Flight Logs:
http://logs.uaventure.com/view/xWtaFRQioDTwiE8j6VDjsG
http://logs.uaventure.com/view/ZPHp2jaXUUbLXZCmzzxGmK

### CODE REVIEW

**Comment by ecmnet** [comment]
@jgoppert Did you compare it to the velocity version? I see a lot of versions for Flow PosHold running around (including my own :-) ) - all seem to have more or less the same result (as far as I've tried them), which leads me to the conclusion that either limits are already reached or, if there is additional potential improvements, they have not been addressed up to now.

**Comment by ecmnet** [comment]
Maybe a consolidated discussion would help here?

**Comment by jgoppert** [comment]
Agreed. I think a square box and hover test would be appropriate.  Maybe 1m x 1m box to start, do 10 laps or so, land where you took off. And for hover. Measure drift over 15 minute flight.

**Comment by jgoppert** [comment]
Did a test and realized my lense was getting condensation from the cold, so heads up to anyone else testing in this weather :)

**Comment by ecmnet** [comment]
That's why I am indoors using a cable-connected quad.. :)

**Comment by LorenzMeier** [comment]
Good to merge?

**Comment by jgoppert** [comment]
Let me flight test indoors right now, then we should be good.

**Comment by jgoppert** [comment]
Pretty impressive indoors as well. 5 minutes, 0 drift. @mhkabir @ecmnet @LorenzMeier  @dkircali

Indoor testing:
https://www.youtube.com/watch?v=N31jdhFWEpY

Flight log:
http://logs.uaventure.com/view/wyuzdHqPd2HfTdjA2WFNSD

**Comment by jgoppert** [comment]
@mhkabir, @ecmnet  Want to test this out  before we merge to master?

**Comment by jgoppert** [comment]
Timeout fixes was already test flown, just forgot to push it up.

**Comment by ecmnet** [comment]
Ok, ich will give it a try right now...

**Comment by jgoppert** [comment]
Probably was a little over confident showing the transmitter 2 inches from the quad in position hold :-P.

**Comment by ecmnet** [comment]
I just took your version and flew it, with not much success I have to admit - so I dogged into your code:

I saw that you compensated flow now not with Z-estimation but with ground distance directly. Thats what I did in my version also. I assume that

float d = _sub_flow.get().ground_distance_m;

gets the sonar results, which doesn't work for me indoors, as I have carpets all around. In my implementation based on mhkabir's version,I take the lidar results to compensate:

float d = _sub_lidar->get().current_distance / ( cosf(_sub_att.get().roll) * cosf(_sub_att.get().pitch));

I also have the effect, that flow is often considered as bad. This reminds me also on the initial implementation of Mohammed, where I had the same effect. Removing the flow quality check and using Lidar gave quite good results.

Refer to https://github.com/PX4/Firmware/pull/3209 for this.

I will try to do the same on your version (takes a bit time).

Further I want to mention, that I noticed, that flow seems not to be direct proportional to the ground distance (at least in my case), which leads to an over-compensation at higher altitudes.

Maybe that helps you a bit.
-Eike

**Comment by jgoppert** [comment]
@ecmnet Do you have skype, or want to jump on mumble. I wold like to debug this with you directly and figure out why you can't reproduce my results. I suspect that the flow quality is the fundamental issue. Why would I never have a flow quality issue on my end in both test videos? Have you checked for black lines in your camera image and focused the camera?

**Comment by jgoppert** [comment]
@ecmnet With regards to distance, it should be:
float d =  -z / ( cosf(_sub_att.get().roll) * cosf(_sub_att.get().pitch));
Where z is the position in the down direction.

If you use lidar, it already is giving you the distance to the image plane, so you should use float d = _sub_lidar->get().current_distance

I think that might have been my fault suggesting that change to @mhkabir when reviewing the code.

**Comment by jgoppert** [comment]
@ecmnet Please also post your log file from your test flight if you can and I will look at it.

**Comment by ecmnet** [comment]
@jgoppert Sry, James, it's quite late in the evening here, so I have to postpone this. I just tried to integrate LIDAR distance, but saw, that you did not subscribe LIDAR separately. How did you distinguish between Sonar and LIDAR?. My PX4Flow is working good as I am able to achieve 10cm accuracy with my modified version.  


... I am just preparing logs for you and the go to sleep :)

**Comment by jgoppert** [comment]
@ecmnet Thanks for the help. The logs should be very informative. Yes, I was forcing it to use the sonar by using: 	float d = _sub_flow.get().ground_distance_m;

**Comment by ecmnet** [comment]
Here are the logs (I did not dig into them, but as you can probably see, I had to take over control massively):

http://logs.uaventure.com/view/J2CaNfT2BwAVaKfRvmM33P

Again you will see massive Flow-Dropouts, I had the same with mhkabir's version, but logging directly quality flow from MAVLink, there were none. Then I removed the quality check and everything was fine.

**Comment by ecmnet** [comment]
Maybe it would be the best to compensate altitude with a kind of sequence:

1. LIDAR, if not
2. SONAR, if not
3. _X(_z) estimation

Of course it would be the best to take an estimated ground distance, that would result of an EKF fusion, not sensor data directly, but NED usually does not give us ground distance.. we would need a stable terrain estimation additionally

**Comment by ecmnet** [comment]
Concerning the factor ( cosf(_sub_att.get().roll) * cosf(_sub_att.get().pitch)); you are right of course...because sensor data is not Z direction, but the distance already..

**Comment by jgoppert** [comment]
@ecmnet Looking at your log the x,y position is barely moving. This is like you said due to the poor flow quality. We need to figure that out, feel free to contact me tomorrow and I will help you track this down.

**Comment by mhkabir** [comment]
Hey James,

Looks OK :) Can you rebase this on lpe_flow and just change out the flow
correction to your method?

**Comment by jgoppert** [comment]
@mhkabir let's do a diff on the branches and see what else we need to bring in, then I will flight test the changes. This already works well indoors and outdoors, so I would rather add your changes as a pr.

**Comment by dkircali** [comment]
@jgoppert I cloned your version and uplaoded to test.
I wanted to try it indoor. However when i switch to position control mode, the command is rejected. 
GPS is connected to board and no LIDAR.
If I disconnect the GPS the board gives a sensor error and I can't arm.
With GPS lock(outdoor) I can switch to position control.
I would appreciate if you could tell me what I am missing to use position control without GPS.

**Comment by jgoppert** [comment]
@dkircali You need to have optical flow quality for it to initialize the position lock. This means you need to  take off in manual/ altitude hold mode, then wait until the main led turns green, then switch to position control mode.

Before you do this, pick up the device without arming it and see if it turns green. Also, make sure you plug in your px4flow and make sure it is in focus at about 1 m, or whatever altitude you want to fly at. Also, make sure there are no odd black lines in your camera etc.

**Comment by jgoppert** [comment]
@mhkabir I cleaned up the branches so we can compare them here: https://github.com/jgoppert/Firmware/compare/lpe-improvements...jgoppert:lpe_flow

I think we should just merge this PR first unless you have any objections. Then we can worry about the rest of your message changes and lidar driver changes as separate PR's.

**Comment by jgoppert** [comment]
@mhkabir I now have a merged version on jgoppert:lpe_flow
I read through your changes and agreed and kept  most.

**Comment by ecmnet** [comment]
@jgoppert I tried again your latest estimator. Did something changed regarding orientation? Compared to mhkabir's implementation, posHold does not work at all for me - whenever I switch on posHold, the vehicle is heading into a (for me random) direction straight away.

**Comment by jgoppert** [comment]
@ecmnet So pos hold works for you now? Do you have a log. I would like to get this merged today. Thanks for reporting the orientation being different from inav/ other lpe branch. I will get that fixed and flight test.

**Comment by ecmnet** [comment]
@jgoppert Yes, it worked. Unfortunately I am currently abroad, so I cannot share any log by now. Regarding the orientation, please double-check that. Last point to mention is that my quad started oscillating when gaining altitude with posHold. I am really curious why.

**Comment by jgoppert** [comment]
@ecmnet @dkircali @mhkabir  I just merged all of mhkabir's work into this. Let's get this tested and get this in master asap. The diff isn't that big but we need to make sure it is all robust. I will be flying this code at a demo in a month so I need it to be very stable and you can count on my support.

**Comment by jgoppert** [comment]
I'm flight testing now, but there is snow here, so I'll mainly be checking flight with GPS outdoors, and indoors with flow.

**Comment by jgoppert** [comment]
I am making some improvements now and will test tomorrow morning. Please wait to test until I push my recent changes. I found an issue with altitude initialization after the merge that I am resolving.

**Comment by ecmnet** [comment]
I would have the chance to perform an indoor test-flight tomorrow.

**Comment by jgoppert** [comment]
OK, this should be good for testing. I haven't been able to test fly it yet though so if you beat me to it, be cautious. @ecmnet @mhkabir @dkircali

**Comment by ecmnet** [comment]
Here you go: http://logs.uaventure.com/view/9UFDkbjkoktZciC7MvDbMA

**Comment by ecmnet** [comment]
And this one: http://logs.uaventure.com/view/jkLpcGgQR94FyjWbzp4KM9

**Comment by jgoppert** [comment]
@ecmnet Logs look good. On the flow distance sensor issue, I think it would be good to compare lidar, vs. sonar before  we choose the default. I realize lidar is more accurate, but am wondering if the alignment and time synchronization from the sonar might compensate and make it better overall?

@ecmnet Regarding why I am using a position based correction, it is because if you fuse the optical flow info into velocity and then integrate it with the accelerometer data etc. it is impossible to hold a position over a long period of time. So to get it work better originally when I wrote it about a year ago I just kept the integrated optical flow as a separate integrator in the estimator, and treated this as a position estimate, which worked much better during my experiments.

The latest commits have really done a lot to clean up initialization and bad data handling so I think this is ready for testing by a wider audience @ecmnet  @mhkabir @dkircali

**Comment by jgoppert** [comment]
Just did a rebase to clean up all of my commits last night.

**Comment by jgoppert** [comment]
Here is the log of the flight over snow: http://logs.uaventure.com/view/BC6UYpDobQQiAfPLk3hF6B

Looks about as good as can be expected over a pretty smooth white snow surface. The only issue I see is that when using sonar for altitude when it goes higher than the sonar can read it doesn't deinitialize the sonar as I would expect.

**Comment by jgoppert** [comment]
OK, another test flight with some minor fixes. Here is a plot of the robustness testing where I tried to trigger as many sensor reinitializatiions as possible to see if the altitude was adversely effected, but it looks good: http://logs.uaventure.com/view/8HWxJ2NJ56HbdTgKKTU9am

Warning while flying this. Current master does not allow you to take of in alt-ctl, so you need to take off in manual, then flip to alt-ctl.

File: src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp [hunk 1]
```diff
@@ -1143,15 +1344,37 @@ void BlockLocalPositionEstimator::correctLidar()
 	// kalman filter correction if no fault
 	if (_lidarFault == FAULT_NONE) {
 		Matrix<float, n_x, n_y_lidar> K = _P * C.transpose() * S_I;
-		_x += K * r;
+		Vector<float, n_x> dx = K * r;
```
Comment (mhkabir) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: Why is this here specifically in correctLidar?

File: src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp [hunk 2]
```diff
@@ -426,29 +515,31 @@ void BlockLocalPositionEstimator::updateHome()
 	_altHome = alt;
 	_gpsAltHome += delta_alt;
 	_baroAltHome +=  delta_alt;
-	_lidarAltHome +=  delta_alt;
-	_sonarAltHome +=  delta_alt;
+	_visionHome(2) += delta_alt;
```
Comment (mhkabir) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: Why was the lidar/sonar offset correction removed?

Comment (jgoppert) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: This correction is basically an offset based on the height of the sensor above the ground. It really makes more sense to be a static parameter. If we try to auto calculate it when the sensor initializes. This leads to unreliable offsets while taking off for sonar and lidar if they are too close to the ground when the vehicle is landed to initialize.

Comment (TSC21) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: So we probably want to keep that correction after a certain height threshold. Is that done already?

File: src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp (hunk 1 above)
Comment (jgoppert) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: The small amount of drift in xy was a result of the correction for sonar/ baro/ lidar. This ensures that if xy can't be measured that the altitude measurements don't inject small amounts of error in xy that eventually add up to  a large value.

File: src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp (hunk 2 above)
Comment (jgoppert) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: No, it is not. Currently it just assumes no offset, which works pretty well with my setup. If you had sonar and lidar both running and they were both offset at different heights, it would lead to some pretty bad oscillations, but that is an extremely unlikley edge case for most users. If we just add a few BlockParamFloats for lidarOffsetZ and sonarOffsetZ, it should be good for all use cases.

**Comment by ecmnet** [comment]
@jgoppert I don't think that having lidar and sonar with different offsets. If you mount both on the same base, lidar and sonar altitude differs by 2-3cm (at least if you use a liar like SF10a).

File: src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp (hunk 2 above)
Comment (ecmnet) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: @jgoppert I don't think that having lidar and sonar with different offsets. If you mount both on the same base, lidar and sonar altitude differs by 2-3cm (at least if you use a liar like SF10a).

Comment (jgoppert) [inline src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp]: @ecmnet 2-3cm difference probably wouldn't be that bad, but I'll go ahead and add the params

**Comment by jgoppert** [comment]
@ecmnet Just added the separate offsets. Let me know how it is working for you.

**Comment by jgoppert** [comment]
@ecmnet @mhkabir @dkircali @tumbili Do you guys have any time to do some testing soon? I will be testing tomorrow and want to try to get this merged with master ASAP, preferably tomorrow.

**Comment by dkircali** [comment]
@jgoppert I am going to test it today but I think I have a problem with my optical flow sensor. There is a black line at the top of the image. But it is visible only when I set video_only to 1.
capture
Maybe that is why I can't initialize the position lock. My main led never turns green. 
I am trying to borrow another one from one of my colleagues.

**Comment by ecmnet** [comment]
@jgoppert I hope I can do a test flight today in the afternoon. Still I find the flow being compensated by sonar altitude only. Although you might be right, with your assumption regarding latency aspects, this will not work for me indoors as sonar is very unstable here (carpets, wall reflexions, etc). So I suggest to use sonar only if it is valid and the fall back to lidar or Z estimation.

**Comment by ecmnet** [comment]
@tumbili Additionally I would like to compare LPE with EKF2 posHold (as shown by Roman in his video) based on flow - is there any public implementation of EKF2 dealing with flow to test ?

**Comment by mhkabir** [comment]
Can you give me a link to the video? :)

Flow fusion for ekf2 is out there somewhere, I've seen a branch sometime
back.

**Comment by ecmnet** [comment]
@mhkabir Here ist is https://www.youtube.com/watch?v=XTKbwJgsuGc&feature=youtu.be. He posted it within the agenda of the last dev call.

**Comment by dkircali** [comment]
@jgoppert @ecmnet  After flashing this branch by "make px4fmu-v2_lpe upload" I can't connect to qgroundcontrol by USB. However I am able to connect with telemetry. I wanted to check the optical flow quality but it is not listed in analyze screen. As I said my main led never becomes green. Moreover now it also rejects altitude hold. 
Do I need to put any txt files to sd card to enable something.

I have also tested the latest official firmware. There is no problem at connecting and altitude control. i can also see the optical flow quality from analyze screen and it is around 250.

Any suggestion is appreciated. Thx

**Comment by dkircali** [comment]
@jgoppert @ecmnet After setting INAV_ENABLED to 0 I am able to get green light. However altitude hold is not reliable. I only use px4flow's sonar. As the altitude change is too much green main light goes blue from green and after 1-2 seconds it becomes green again. But the altitude hold was too unstable that my quad climbed too fast and hit the ceiling.
By the way how do you mount the optical flow sensor to the frame. I thought vibration is an important issue so added some foam to dampen the vibrations. Here is a picture of my setup:

**Comment by jgoppert** [comment]
@dkircali Can you give me a flight log? I was flying with sonar only and didn't have those problems, so I want to see if you have found and issue or it is due to your optical flow unit.

@ecmnet I was doing sonar flight over carpet and didn't have a problem, can you give me a flight log showing the issue. Also, I'm switching to prefer lidar if plugged in right now, so you should be good to go.

**Comment by jgoppert** [comment]
I'm flight testing now with sonar, will let you know how it turns out.

**Comment by ecmnet** [comment]
@jgoppert Just had a short test flight. Light conditions were not optimal so I had some flow dropouts, but here's the log. Manual correction of the position seems to be very snappy (I think that is not an issue of LPE, but of mc_pos_control). Anyway, here is the log: 

http://logs.uaventure.com/view/eX3TswUBxGwBnZ2WC5uhqd

**Comment by ecmnet** [comment]
@dkircali Did you experience a kind of 'jump' when switching to altHold? I had this issue with mc_pos_control, which is solved by https://github.com/PX4/Firmware/pull/3822. To confirm that LPE is working properly just check the Local NED-Z when holding the vehicle above ground.

**Comment by jgoppert** [comment]
I just had a great flight test!

https://www.youtube.com/watch?v=w4mOagDO2M0

http://logs.uaventure.com/view/j7HnB9uzyxBv3dD4x4mkrj
I tested sonar dist max etc. Also it was working so well I started feeding in position and altitude commands and looking at the step responses. I agree it is a bit too snappy. Somehow the pos ctrl gains must have been changed. Did anything like this happen recently to mc_pos_control?

**Comment by ecmnet** [comment]
@jgoppert Just did a test with your latest changes. In low light conditions still quite reasonable results. For mc_pos_control, I am already using  https://github.com/PX4/Firmware/pull/3822. Seems a lot more snappier than the master implementation. Still I have the sonar/lidar fusion issue.

**Comment by jgoppert** [comment]
Here are my straight pos hold results. Had a fly off at the end, this was before the 1cf7 patch. Just realized it was hard to see how awesome the altitue and position hold are now with me moving the stick around so much in the other video.

**Comment by jgoppert** [comment]
@ecmnet You should see the lidar out of range message when you cover up the lidar. Does this happen? This should cause a lidar fault and timeout. Sonar will correct if either lidar is not initialized or has a fault.

**Comment by jgoppert** [comment]
@LorenzMeier can you can a look at this and tell me what else you would need to see before we merge. I think this is significantly better than what is on master. If you want me to split out the lidar driver changes or sdlog2 changes, let me know, I will do that.

**Comment by ecmnet** [comment]
@jgoppert Yes! It works (1cf7 patch). Covering Lidar returns liar timeout and Z-Position is still estimated using sonar. I am just preparing a flight video.

**Comment by jgoppert** [comment]
@ecmnet Try this latest change, it should do what you want. It keeps sensors initialized as long as they have valid readings. I have a meeting so can't flight test more today.

**Comment by ecmnet** [comment]
@jgoppert Last flight for today (15 mins). Not much to mention about this. Everything we discussed seem to work. One thing to mention: Below 0.3m posHold did not work anymore. Got a message 'sonar min distance' and saw a heavy drift. I hope I have time tomorrow to do some flights with better light condition.

**Comment by ecmnet** [comment]
@jgoppert  @tumbili  Next step for me would be a kind of benchmark against EKF2 flow fusion.

**Comment by jgoppert** [comment]
@ecmnet thanks for all of the testing today!

**Comment by ecmnet** [comment]
@jgoppert Really the last thing for today 😄: Discussing the auto-landing feature with @sanderux it turned out that there are several requirements to determine the landing speed ( https://github.com/PX4/Firmware/pull/3898#discussion_r54687422). I think LPE has to provide that info as well. I hope I can check that by tomorrow.

**Comment by sanderux** [comment]
It doesnt seem to be very picky on this: https://github.com/jgoppert/Firmware/blob/master/src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp#L809 :)

**Comment by ecmnet** [comment]
@sanderux Thanks for the hint. I haven't had a look into the code up to now.

**Comment by ecmnet** [comment]
@mhkabir It makes generally no sense to restrict the implementation to certain assumptions - even if they cover 90% of the use-cases, when there is a simple and general solution possible, which covers 100% of the cases. And that is, what James has implemented now and it is working. Besides that, it always makes sense to use all available information, choosing the best one in respect to the current situation ( that is why - generally spoken - sensor fusion makes sense).

Regarding the range finder: Accuracy is not the only criteria for choosing a range finder. SF10a has no single dropout, works perfectly on all type of grounds, etc. And with the successor of Sf10a the range is extended to 100m altitude at a reasonable price.

**Comment by mhkabir** [comment]
@ecmnet Why don't you send in a PR against this branch with your changes? 8)

**Comment by mhkabir** [comment]
@sanderux The bottom distance thing is on our list of things to fix once we have optical flow working perfect. We'll be separating out terrain altitude as a state, and adding a little something to mc_pos_control to do the terrain following.

**Comment by ecmnet** [comment]
@mhkabir James already implemented it in https://github.com/jgoppert/Firmware/commit/5b0006974c38010000f0df89bc7d314ade797c93

**Comment by mhkabir** [comment]
@ecmnet Ah, nice :)

**Comment by ecmnet** [comment]
@mhkabir I am on the way to do this :)

**Comment by ecmnet** [comment]
Just to prove this (Red = raw lidar, Yellow = raw sonar)

chart

I've just implemented a preliminary simple solution in my branch. It currently depends strongly on the assumption that offsets are compensated by setting the parameters accordingly. I am just thinking of a mechanism to determine the relative offset of both automatically.

**Comment by jgoppert** [comment]
@mhkabir @sanderux @ecmnet @dkircali  This implements terrain estimation and terrain following now. Looking forward to testing out tomorrow. I've done some bench top tests and it looks good.

**Comment by ecmnet** [comment]
@jgoppert Will give it a testflight on friday.

**Comment by mhkabir** [comment]
@jgoppert 
Hey James, looking good! I believe we still need to do stuff in the controllers to actually make it work?

What I thought was : 
* Follow ground when terrain estimate available.
* When terrain estimate unavailable, then switch to Local Z, but reset the Z setpoint accordingly, and in a continuous manner such that we have no jumps.

**Comment by jgoppert** [comment]
@mhkabir I've already got it working so that there is a smooth transition with or without sonar/lidar. I even did a flight on just baro to make sure it still works well. So no rush to fix the controller. I think that the other estimators are assuming the same. But I do think we should fix that eventually so that it is switchable between AGL and ASL altitude hold.

baro  + gps only:
http://logs.uaventure.com/view/SXH42zHBQSf5kgFUVDYjWH

sonar + gps + flow (but it was dark so it didn't initialize often):
http://logs.uaventure.com/view/WViSbzdRY2NW4jPS74ca4h

Hopefully I will get lidar + gps + flow and lidar + flow tested tomorrow when I get some light :-).

**Comment by jgoppert** [comment]
The terrain state (tz) is the relative altitude of the terrain to the home position. So 1 means it is one meter below the home altitude. The dynamics are just  d/dt(tz) = 0;

The distance sensors then measure:
y = - (z - tz)

I also just fixed some issues with GPS once I enabled the full blown fault detection. The fault detection helped me figure out that our GPS altitude had a wrong sign in the difference from the home altitude :-P. Also, when you arm it was throwing a GPS fault, because it was trying to shift local x and creating velocities that GPS disagreed with, I fixed this by reinitializing the covariance matrix when the home location is reset.

**Comment by jgoppert** [comment]
@LorenzMeier @tumbili @mhkabir  Plan to merge this with master after the build passes as discussed on the dev call. The diff is down to just LPE files so all changes should be isolated form the rest of the code.

**Comment by rousselmanu** [comment]
@dkircali I have the same upper black line as in your image when VIDEO_ONLY=1 (tried on my 2 px4flow, same result), but position hold/control is still working well with master firmware and LPE build.

**Comment by MegaAdragon** [comment]
@jgoppert @mhkabir @ecmnet I am working on an indoor application and tried using this branch. I am interested in a stable position control.

I am using the sonar from the PXFlow and no gps (because I'm indoor).
This is my first project with the Pixhawk and PXFlow and I am trying to optimize my setup. 

In this test flight I used the altitude control, but the drone wasn't really stable. The altitude permanently jumped and returned to the initial value.
http://logs.uaventure.com/view/PHTeUb9odKaXDybiuq9EV6

After switching to position control the drone drifted to one side and I tried to correct it with the RC. This led to the drone making an uncontrollable move, it flipped and crashed.

Do you have any input in how I can make the altitude hold more stable and what could be the fault with position hold. 
The ground in this room has a flat surface, so there should be no problem with the sonar.
My biggest problem is the fact that I don't have much space indoor. So I have to somehow manually stop the drone from hitting something if it drifts off too far.

**Comment by jgoppert** [comment]
@MegaAdragon Thanks for testing. Please use this branch, https://github.com/PX4/Firmware/pull/4036, It has fixes for the issues you were experiencing with altitude. If you look at your altitude, the sonar misreads occuring every second or so are causing large altitude spikes. I enabled fault detection in the most recent version of the code, where the sonar weighting is decreased when a fault is detected.

If position control fails to hold the vehicle in one position you probably need to update the firmware and focus the camera on your flow. You can do this by plugging the flow into qgroundccontrol to update the firmare, or do it manually vis the git repo and typing make usb-upload. To focus the camera, using the latest firmware for flow and qgc you should get a camera view in qgc of the flow video. Hold the flow about 2m above text and focus the camera on the text by loosening the set screw and turning the lense until it is in focus.
